[Jiang Zhimeng] iP

asaierika opened the pull request:
### Rabbit

> “Why not do me a favour, close the window and free both of us?" -Rabbit

Let's admit it - who likes taking notes? I'm sure you don't, and same goes for Rabbit. ~~But she can't complain about it~~ - even though she does, a lot - since it's her job to help you keep track of tasks. Rabbit is 

- text-based
- easy to learn
- cute

All you need to do is,

1. download it 
2. double-click it.
3. read the instructions.
4. start using despite Rabbit's complaint.

What you can do:

- [ ] Keep track of the tasks.
- [ ] Change Rabbit's look if you are really bored.
- [ ] Get scolded by Rabbit when you get the format wrong.:+1:
[https://github.com/asaierika/ip/releases/tag/A-Release](url)
Here's the `main` method:
```Java
  @Override
    public void start(Stage stage) {
        FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));

        try {
            rabbit = new Rabbit();
            AnchorPane ap = fxmlLoader.load();
            Scene scene = new Scene(ap);
            stage.setScene(scene);
            fxmlLoader.<MainWindow>getController().setRabbit(rabbit);
            stage.show();
            fxmlLoader.<MainWindow>getController().showText(rabbit.GREET);
            fxmlLoader.<MainWindow>getController().showText(rabbit.INSTRUCTIONS);
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ImportDataException e) {
            fxmlLoader.<MainWindow>getController().showText(e.toString());
        }
    }
}
```

asaierika committed:
Add Rabbit.java which is the class of the robot Rabbit.

asaierika committed:
Add function AddToList() to Rabbit.java where Rabbit keeps the user's input in a list.

asaierika committed:
Deleled Duke.java
Added Task.java as a class of task noted down by Rabbit
Added functions list(), mark() and unmark()
Changed addToList() and main()

asaierika committed:
Added derived classes of Task: Todo, Deadline and Event.
Added functions in Rabbit to create tasks of different types.
Changed task to abstract

asaierika committed:
Fixed bugs in unmark and main function in Rabbit
Changed input.txt EXPECTED.txt and runtest.sh to test Rabbit

asaierika committed:
Added exception classes: RabbitException, MarkUnmarkException and AddToListException
Changed mark, unmark, addToList and main methods in Rabbit to handle exceptions

asaierika committed:
Added DeleteException clsss
Added delete method in Rabbit to enable deleting a task
Changed input and EXPECTED to test delete method

asaierika committed:
Change formatting of the code

asaierika committed:
Add save feature so that the list is kept in ./data/data.txt whenever there is a change in the lsit.

asaierika committed:
Change the implementation of the time of Task from String to LocalDataTime.

asaierika committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Rabbit.java

asaierika committed:
Add Parser.java, Storage.java, TaskList.java, Ui.java. Change methods in Rabbit.java.

asaierika committed:
Add all files into packages:Rabbit, Rabbit.RabbitException, Rabbit.util and Rabbit.Task.

asaierika committed:
Add Gradle support.

asaierika committed:
Add unit tests ParserTest and TaskListTest to test Parser and TaskList respectively.

asaierika committed:
Add javadoc to some public methods.

asaierika committed:
Change codes style according to coding standard.

asaierika committed:
Add Find method in TaskList to enable the user to find a task with keywords.

asaierika committed:
Merge branch 'branch-A-CodingStandard'

asaierika committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/Rabbit/util/Parser.java
#	src/main/java/Rabbit/util/TaskList.java
#	src/main/java/Rabbit/util/Ui.java

tlx02 commented on src/main/java/Rabbit/util/Ui.java:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Constructor of Ui.
+     */
+    public Ui(){
```
Might be good to include a white space between ')' and '{'.

tlx02 commented on src/main/java/Rabbit/util/Ui.java:
```diff
@@ -0,0 +1,126 @@
+     */
+    public Ui(){
+        this.sc = new Scanner(System.in);
+    };
```
I don't think ';' is needed here?

tlx02 commented on src/test/java/Rabbit/util/ParserTest.java:
```diff
@@ -0,0 +1,24 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+import Rabbit.RabbitException.RabbitException;
+import java.time.LocalDateTime;
```
Are these statements in the correct order?

tlx02 commented on src/test/java/Rabbit/util/TaskListTest.java:
```diff
@@ -0,0 +1,24 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+import Rabbit.RabbitException.RabbitException;
+import java.time.LocalDateTime;
```
Are these in the correct order? Might want to check for other classes too.

Aishwarya-Hariharan-Iyer commented on src/main/java/Rabbit/util/Parser.java:
```diff
@@ -0,0 +1,224 @@
+     * @throws ImportDataException
+     */
+    public static Task parseImport(String input) throws ImportDataException {
+        char type = input.charAt(0);
```
I really like how you catch errors in this method! Do you think that perhaps Java String [split method with regex ](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#split(java.lang.String)) instead of using `substring` might help make the method briefer?

Aishwarya-Hariharan-Iyer commented on src/main/java/Rabbit/util/TaskList.java:
```diff
@@ -0,0 +1,203 @@
+                    continue;
+                }
+
+                for (int j = 0; j < content.length() - keyword.length() + 1; j++) {
```
Perhaps you could consider adding a method in the Task class/subclasses to check if they contain a certain word? That way, it might add to OOP and code quality?

Aishwarya-Hariharan-Iyer commented on src/main/java/Rabbit/util/Parser.java:
```diff
@@ -0,0 +1,224 @@
+        }
+
+        if (list.get(i - 1).isDone()) {
+            throw new MarkUnmarkException(MarkUnmarkException.Type.MARKREPEAT);
```
I find this really well done! I feel that this exception is not one that we commonly think of so adding it was very important and well thought-out!

Aishwarya-Hariharan-Iyer commented on src/main/java/Rabbit/util/TaskList.java:
```diff
@@ -0,0 +1,203 @@
+    public String addToList(TaskType task, String input) throws AddToListException {
+        if (this.list.size() == 100) {
+            // throws an exception when there are already 100 lines in
+            // the list when the user is trying to input a new line.
```
I find your JavaDoc descriptions and in-code comments really useful in navigating the purpose of methods and fields! It seems to adhere very well to the Guideline of **"Comment minimally, but sufficiently"** given in the [Code Quality Guidelines](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html).

Aishwarya-Hariharan-Iyer commented on src/main/java/Rabbit/RabbitException/FindFormatException.java:
```diff
@@ -0,0 +1,8 @@
+public class FindFormatException extends RabbitException {
+    @Override
+    public String toString() {
+        return "Hmmm...I don't get what you're trying to find. Human language please?";
```
I really like how comprehensive all your Exceptions are! A small consideration to make: perhaps, adding more details to the printed message here might help highlight the difference between what the user must do after getting a `FindFormatException` and an `InvalidInputException`.

Aishwarya-Hariharan-Iyer reviewed (commented):
LGTM. Just a few observations and suggestions with respect to Code Quality.

tlx02 commented on src/main/java/Rabbit/RabbitException/AddToListException.java:
```diff
@@ -0,0 +1,31 @@
+        switch(this.type) {
+        case FULL:
+            return "There are too many tasks in the list!\n"
+                    + "Don't exceed 100 please. The poor notebook is full.";
```
Nice line break here!

tlx02 commented on src/main/java/Rabbit/RabbitException/DeleteException.java:
```diff
@@ -0,0 +1,24 @@
+        switch(this.type) {
+            case FORMAT:
+                return "Type 'delete index' to delete a task if that's what you want.";
+            case INDEX:
```
Your switch statements for AddToListException were in the right format. Perhaps you missed out this part while editing?

tlx02 commented on src/main/java/Rabbit/util/Parser.java:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    /**
+     * Returning the first word of the string.
```
Might want to use 'Returns' instead

tlx02 commented on src/main/java/Rabbit/util/Storage.java:
```diff
@@ -0,0 +1,108 @@
+     */
+    public void initiate() {
+        // if directory ./date does not exist, create it
+            try {
+                Files.createDirectories(path.getParent());
+            } catch (IOException e) {
+                // directory already exists
+                return;
+            }
```
Not sure if the indentation here is correct.

tlx02 commented on src/main/java/Rabbit/util/Parser.java:
```diff
@@ -0,0 +1,224 @@
+     * @return the matching tasks.
+     * @throws FindFormatException
+     */
+    public static String parseFind(String input, TaskList list) throws FindFormatException {
```
I like how your methods are all clearly and correctly named as verbs!

tlx02 reviewed (commented):
Looks good to merge. Just a few minor coding standard violations.

asaierika committed:
Add GUI to Rabbit using javaFX.
Delete Ui.java as old UI system is no longer needed.

asaierika committed:
Add Assertions to TaskList class

Assertions are good to detect bugs.

Let's add assertions in delete, mark and unmark methods in TaskList to
ensure that the task to be operated on is not null. Even though the methods
that are calling these methods throw exceptions when the task to be operated
is null, it is safer to use assertions as future modifications may result
in bugs that cannot be detected by the old method.

asaierika committed:
Parse: use String.split() to replace old method

Old methods in Parse split the string using for loop which is slow and
cumbersome.

Let's use String.split() is used to make the methods shorter.

Additional change package name according to naming convention.

asaierika committed:
Merge pull request #2 from asaierika/branch-A-Assertions

Add Assertions to TaskList class

asaierika committed:
Merge commit '750cb0ffe35b0387752cf6bfedbe30164dfd8c6d' into branch-A-CodeQuality

# Conflicts:
#	data/data.txt
#	src/main/java/Rabbit/util/Parser.java
#	src/main/java/Rabbit/util/TaskList.java

asaierika committed:
Merge pull request #3 from asaierika/branch-A-CodeQuality

Branch a code quality

asaierika committed:
Add methods to edit a task content and time

It is useful to be able to change a task's content and time without
deleting and recreating it.

Let's add edit() method in TaskList and parseEdit() in Parse to
implement this new feature.

asaierika committed:
Merge pull request #4 from asaierika/branch-BCD-Extension

Add methods to edit a task content and time

asaierika committed:
Improve the GUI

The GUI needs to be improved to look better.

Let's clip the avatar into a circle and change the colour of the box.

asaierika committed:
Add User Guide in README.md

asaierika committed:
Edit User Guid in README.md

asaierika committed:
Fix bugs on importing fxml

asaierika committed:
Edit README.md

asaierika committed:
Update README.md

asaierika committed:
Update README.md

asaierika committed:
Modify README.md
